File: javascript/duplicates/containsNearbyDuplicate.js
// Example: Check if the array contains duplicates within k distance
// Duplicates make this problem harder because you must track indices and check the distance between duplicates

function containsNearbyDuplicate(arr, k) {
  const seen = new Map();
  for (let i = 0; i < arr.length; i++) {
    if (seen.has(arr[i]) && i - seen.get(arr[i]) <= k) {
      return true;
    }
    seen.set(arr[i], i);
  }
  return false;
}

const arr = [1, 2, 3, 1, 4, 5, 1];
const k = 3;
console.log(
  "Contains nearby duplicate within",
  k,
  ":",
  containsNearbyDuplicate(arr, k)
); // Output: true

File: python/arrays/contains_nearby_duplicate.py
# Example: Check if the list contains duplicates within k distance
# Duplicates make this problem harder because you must track indices and check the distance between duplicates

def contains_nearby_duplicate(arr, k):
    seen = {}
    for i, num in enumerate(arr):
        if num in seen and i - seen[num] <= k:
            return True
        seen[num] = i
    return False

arr = [1, 2, 3, 1, 4, 5, 1]
k = 3
print('Contains nearby duplicate within', k, ':', contains_nearby_duplicate(arr, k))  # Output: True 